posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md

```diff
@@ -0,0 +1,120 @@
+---
+layout: post
+title: "Using rustc_codegen_cranelift for debug builds"
+author: Joshua Nelson
+team: The Compiler Team <https://www.rust-lang.org/governance/teams/compiler>
+---
+
+## What is `rustc_codegen_cranelift`?
+
+[`rustc_codegen_cranelift`], or just `cg_clif` for short, is a new experimental
+codegen backend for the Rust compiler. The existing backend is LLVM, which very
+good at producing fast, highly optimized code, but is not very good at
+compiling code quickly. `cg_clif`, which uses the [Cranelift] project, would
+provide a fast backend which greatly improves compile times, at the cost of
+only performing very few optimizations. This is a great fit for debug builds, and the hope is
+that `cg_clif` will eventually be the default backend in debug mode.
+
+## What is the progress of using `rustc_codegen_cranelift` for debug builds?
+
+There has been an [Major Change Proposal][MCP] open for some time for making
+`cg_clif` part of the main Rust repository. Recently, [the MCP was
+accepted][compiler-team#270] and the compiler team [merged][#77975]
+`rustc_cranelift_codegen` [into the main Rust git repository][#77975].
+`cg_clif` still is not distributed with `rustup`, but this means you can now
+build it from source in-tree!
+
+## How do I use `rustc_codegen_cranelift`?
+
+In this section, I'll walk through step-by-step how to build the new backend from source, then use it on your own projects. All code is copy/paste-able, and each step is explained.
+
+First, let's build `cg_clif` from source.
+
+```sh
+$ git clone https://github.com/bjorn3/rustc_codegen_cranelift.git
+$ ./prepare.sh
+$ cargo build --release
+$ ./build_sysroot/build_sysroot.sh
+```
+
+Now, we can start using it for our own projects. For demonstration
+purposes, I'll be be using `cargo`, but you can use any Rust project supported
+by `cg_clif`.
+
+```
+$ cd ..
+$ git clone https://github.com/rust-lang/cargo/
+$ cd cargo
+$ ../cargo.sh build
+...
+    Finished dev [unoptimized + debuginfo] target(s) in 49.93s
+```
+
+It works! For comparison, let's see how long the equivalent LLVM backend would
+take.
+
+```sh
+$ rustup install nightly-2020-10-31
+$ cargo +nightly-2020-10-31 build
+...
+    Finished dev [unoptimized + debuginfo] target(s) in 54.64s
+```
+
+LLVM takes a full 5 seconds longer for a full build.
```

Comment:
How much is the difference in incremental recompilation time? cg_clif doesn't support internal parallelism until parallel rustc lands. This means that cg_llvm can normally utilize the cpu cores more effectively when there are few crates that can be built at a given time. In incremental mode, it likely only needs to recompile a single cgu, so there won't be any internal parallelism for cg_llvm too.

Reply:
Does `cg_llvm` support internal parallelism? Is that parallism in LLVM itself?

Reply:
cg_clif was actually slower, not faster :/

Reply:
cg_llvm optimizes codegen units in parallel.

Reply:
Still slower if I only change main.rs:
```
$ CARGO_TARGET_DIR=cg_clif ../rustc_codegen_cranelift/cargo.sh build
    Finished dev [unoptimized + debuginfo] target(s) in 0.89s
$ cargo +nightly-2020-10-31 build
   Compiling cargo v0.50.0 (/home/joshua/cargo)
    Finished dev [unoptimized + debuginfo] target(s) in 0.89s
```
Maybe it's because I compiled cg_clif locally but the nightly version comes from CI and does LTO?

Reply:
serde_derive shows up a lot in the profile. I think it is simply due to the fact that serde_derive is not very optimized in this case.

Reply:
Probably https://github.com/serde-rs/json/issues/727 then.

Reply:
That issue is about compiletime of crates using serde_json due to large amounts of ir being generated. This issue is serde_derive being slow. It takes ~14% of the compilation time for libcargo.

Reply:
For cg_llvm it is <3% of the compilation time spent in serde_derive for libcargo.